Have local() set the SHELL env var for consistent behavior.

joshourisman opened the pull request:
As show in #1661, the fact that `local()` (mistakenly stated as `prefix` in that issue) leaves the value of `$SHELL` unchanged can cause issues if the command being run depends on `$SHELL` referencing the actual shell being used. This pull request modifies `local()` to ensure that `$SHELL` is set appropriately. A test is included to ensure that `$SHELL` matches `$0` when the `local()` command is executed.

ploxiln commented on fabric/operations.py:
```diff
@@ -1198,8 +1198,14 @@ def local(command, capture=False, shell=None):
+    if set_shell is None:
+        del env.shell_env['SHELL']
+    else:
+        env.shell_env['SHELL'] = set_shell
```
This is messy and unnecessary. Just copy the env, modify it, and pass it to `Popen()`:

```python
sub_env = os.environ.copy()
sub_env['SHELL'] = shell or '/bin/sh'
...
subprocess.Popen(cmd_arg, env=sub_env, shell=True, ...)
```

I still think just deleting SHELL from `sub_env` would be simpler, as it gets the actual shell to set the appropriate value, as bash does. I don't know if fabric really runs on windows (I think it theoretically does), but that would be an example of a case where `Popen()` would not actually use /bin/sh so assuming it does here would be wrong.

joshourisman replied:
Definitely unnecessarily messy, not sure what I was thinking. Will fix that.

I tried the approach of deleting SHELL from sub_env, but that doesn't actually seem to solve the problem. In my case, `local('echo $SHELL')` still returns `/usr/local/bin/fish`.

Windows not working with this is a good point.

ploxiln replied:
... might be because you tried with `env.shell_env` which is an additive-only fabric thing.

Anyway, thanks for the report and the PR, it's a good find :)

ploxiln commented on tests/test_context_managers.py:
```diff
@@ -125,6 +125,19 @@ def test_cd_prefix():
+
+def test_prefix_shell_variable():
+    """
+    prefix should set the $SHELL varible for consistent behavior
```
this whole issue really has nothing at all to do with `prefix()`

joshourisman replied:
Oops, you're right. Forgot to move that. :(

joshourisman commented:
Fixed the test, test_operations.py instead of test_context_managers.py, and doesn't use or reference prefix.

I also cleaned up the code actually setting the SHELL env var using the approach @ploxiln recommended. I tried passing in a clean env with no SHELL var set at all, but my local $SHELL value still came through.

joshourisman commented:
Oh, also just skipping modifications to env for win32. I don't currently have access to a Windows machine on which I could test this out, so this seems like the safest approach.

ploxiln commented:
Seems reasonable to me, I'd be fine with this.

But deleting SHELL from env works for me:

```
$ SHELL=fish python -c 'from subprocess import Popen; Popen("echo SHELL=$SHELL", shell=True)'
SHELL=fish

$ cat test.py 
import subprocess
import os

sub_env = os.environ.copy()
if 'SHELL' in sub_env:
    del sub_env['SHELL']

subprocess.Popen("echo SHELL=$SHELL", env=sub_env, shell=True)

$ SHELL=fish python test.py
SHELL=/bin/bash
```

joshourisman commented:
Interesting. If I run that exact same test.py, i get this:

```shell
$ python2 test.py 
SHELL=/usr/local/bin/fish
```

I'm running fish 2.6.0 and Python 2.7.13.

joshourisman commented:
Ah ha! It's apparently because you're explicitly setting SHELL when you run it. If I do that, I get the same result as you:

```python
$ SHELL=fish python2 test.py 
$ 
```

joshourisman commented:
Ok, nevermind. I have no idea what's going on.

```shell
$ bash
bash-3.2$ SHELL=fish python2 test.py
bash-3.2$ SHELL=/usr/local/bin/fish
```

ploxiln commented:
hmm could be `/etc/profile` or `~/.bashrc` or something ¯\\_(ツ)_/¯
